modules/codegen/src/main/scala/com/twilio/guardrail/generators/EndpointsClientGenerator.scala

```diff
@@ -0,0 +1,462 @@
+              case param"$_: org.scalajs.dom.raw.File = $_"         => liftFileTerm _
+              case param"$_: Option[$_]"                            => liftOptionTerm _
+              case param"$_: Option[$_] = $_"                       => liftOptionTerm _
+              case _                                                => liftTerm _
+            }
+
+            val args: List[Term] = parameters.map {
+              case ScalaParameter(_, param, paramName, argName, _) =>
+                lifter(param)(paramName, argName)
+            }
+            Some(q"List[Option[(String, Either[String, org.scalajs.dom.raw.File])]](..${args}).flatten")
+          } else {
+            def liftTerm(tParamName: Term, tName: RawParameterName) =
+              q"List((${tName.toLit}, Formatter.show($tParamName)))"
+
+            def liftIterable(tParamName: Term, tName: RawParameterName) =
+              q"$tParamName.toList.map((${tName.toLit}, _))"
+
+            def liftOptionTerm(tpe: Type)(tParamName: Term, tName: RawParameterName) = {
+              val lifter = tpe match {
+                case t"Iterable[$_]" => liftIterable _
+                case _               => liftTerm _
+              }
+              q"${tParamName}.toList.flatMap(${Term.Block(List(q" x => ${lifter(Term.Name("x"), tName)}"))})"
+            }
+
+            val args: List[Term] = parameters.foldLeft(List.empty[Term]) {
+              case (a, ScalaParameter(_, param, paramName, argName, _)) =>
+                val lifter: (Term.Name, RawParameterName) => Term =
+                  param match {
+                    case param"$_: Option[$tpe]"      => liftOptionTerm(tpe) _
+                    case param"$_: Option[$tpe] = $_" => liftOptionTerm(tpe) _
+                    case _                            => liftTerm _
+                  }
+                a :+ lifter(paramName, argName)
+            }
+            Some(q"List(..$args).flatten")
+          }
+
+        def generateHeaderParams(parameters: List[ScalaParameter[ScalaLanguage]]): Term = {
+          def liftOptionTerm(tParamName: Term.Name, tName: RawParameterName) =
+            q"$tParamName.map(v => RawHeader(${tName.toLit}, Formatter.show(v)))"
+
+          def liftTerm(tParamName: Term.Name, tName: RawParameterName) =
+            q"Some(RawHeader(${tName.toLit}, Formatter.show($tParamName)))"
+
+          val args: List[Term] = parameters.foldLeft(List.empty[Term]) {
+            case (a, ScalaParameter(_, param, paramName, argName, _)) =>
+              val lifter: (Term.Name, RawParameterName) => Term = param match {
+                case param"$_: Option[$_]"      => liftOptionTerm _
+                case param"$_: Option[$_] = $_" => liftOptionTerm _
+                case _                          => liftTerm _
+              }
+              a :+ lifter(paramName, argName)
+          }
+          q"scala.collection.immutable.Seq[Option[String]](..$args).flatten"
+        }
```

Comment:
A lot of the form data and header param generation code looks _really_ similar to the scala/akka code.  Is this worth parameterizing?

Reply:
Possibly. I'm trying to avoid creating more spaghetti, especially as new frameworks and languages come into the mix.